### PR TITLE
feat: make DB catalog work w/ transaction aborts

### PR DESCRIPTION
**Pull request by crepererum**
# Abstract
This allows us to handle transaction aborts (e.g. when the write IO for the transaction file fails) correctly. Before this change aborts would lead to an inconsistent in-memory VS persisted state (in-memory says it's persisted while in fact the transaction wasn't written).

# Background
The reason this problem existed was some design mistake and laziness. When I designed the persisted catalog I somewhat assumed that we could just copy the whole managed in-memory state at transaction start, modify it during the transaction and commit it in the end. However this doesn't work for the real DB, so during the integration I made the "copy" step optional. Turns out that breaks the abort semantic :sweat_smile: 

# Fix
So the fix comes basically in 3 steps:

1. Add a generic test for `CatalogState` implementations that tries to check some invariants/features (incl. what happens during aborts). This one passes for test catalog but fails for the real one.
2. Rework the `CatalogState` interface to be more generic on how an implementation wants to handle state changes (allowing for the simple "copy everything", delayed action which is used by the DB now, as well as a rollback mechanic should we ever need this).
3. Fix the real DB catalog to work with the new interface and also to account for aborts correctly. Now the test added in (1) passes.

# References
This should help w/ #1725.

### CODE REVIEW

File: server/src/db.rs [hunk 1]
```diff
@@ -1141,40 +1265,131 @@ impl CatalogState for Catalog {
...
+        ) {
+            (false, Some(chunk)) => {
                 // Chunk exists => should be in frozen stage and will transition from there
 
                 // Relock the chunk again (nothing else should have been able
                 // to modify the chunk state while we were moving it
-                let mut chunk = chunk.write();
+                let chunk = chunk.read();
+
+                // check if chunk already exists
```
Comment (tustvold) [inline server/src/db.rs:1289]: I can't help feeling these should be assertions, something has gone very wrong if we've gotten to this point and these don't hold?

File: server/src/db.rs [hunk 2]
```diff
@@ -1095,24 +1102,139 @@ pub struct CatalogEmptyInput {
...
+                        table_name,
+                        partition_key,
+                        chunk_id,
+                        inner,
+                    } => {
+                        let partition = catalog
+                            .get_or_create_partition(table_name.clone(), partition_key.clone());
+                        let partition = partition.read();
+
+                        match partition.chunk(chunk_id) {
```
Comment (tustvold) [inline server/src/db.rs:1207]: Similar to the below, I can't help feeling this code should assert - if it fails something has gone spectacularly wrong

File: server/src/db.rs [hunk 3]
```diff
@@ -1141,40 +1265,131 @@ impl CatalogState for Catalog {
...
+        tstate: &mut Self::TransactionState,
+        path: DirsAndFileName,
+    ) -> parquet_file::catalog::Result<()> {
+        if tstate.removed_files.contains(&path) {
+            return Err(parquet_file::catalog::Error::ParquetFileDoesNotExist { path });
+        }
+
+        let mut actions: Vec<TransactionCommitAction> = vec![];
+
+        for partition in tstate.catalog.partitions() {
```
Comment (tustvold) [inline server/src/db.rs:1360]: I can't help feeling it should be possible to identify the partition from the path, but... :man_shrugging:

**Review by tustvold — APPROVED** [review]
LGTM. Having the more likely to fail operation occur before the operation that should be infallible makes sense to me :+1:

File: server/src/db.rs (hunk 1 above)
Comment (crepererum) [reply server/src/db.rs:1289]: This could be hit during replay (in case a catalog adds a file twice in subsequent transactions), so this shouldn't panic. It should return an error so we could report the catalog as broken w/ some human-readable error.

**Review by alamb — APPROVED** [review]
This code makes sense to me, though my gut feeling is that the transaction handling code is more complicated than the problem it is currently solving. I don't have any concrete suggestions at the moment -- but it might be worth taking a step back (maybe with data about how the catalog is working on IOx in tools) and seeing if we could simplify this somehow

Right now there appears to be somewhat of a circular dependency:

```
db --> catalog::TransactionHandle --> db::TransactionState
```

So `db` depends on `catalog` which then depends on `db` again (though I realize this is done via traits).

Would it be possible to make the hierarchy more like:

```
db --> db::TransactionState --> catalog::TransactionHandle 
```

In other words have TransactionState own the catalog transaction handle rather than the other way around?

**Comment by crepererum** [comment]
@alamb let's go ahead with this PR here, I'll tackle #1740 afterwards which will simplify things a bit.
